File: repomap_roles/priority_engine.py
"""
Priority assignment and bucketing engine.

Converts role scores into priority buckets:
- PRIMARY (0.7 – 1.0): Must understand
- SUPPORTING (0.4 – 0.7): Related logic
- CONTEXT (0.1 – 0.4): Additional context
"""

from typing import Dict, List, Set, Tuple
from dataclasses import dataclass

from config import Role, PRIORITY_THRESHOLDS, MAX_FILES_PER_PRIORITY


@dataclass
class PrioritizedFile:
    """Represents a file with priority and relevance metadata."""
    file_path: str
    score: float
    priority: str  # "primary", "supporting", "context"
    roles: Set[Role]
    reason: str = ""
    confidence: float = 1.0
    
    def to_dict(self) -> dict:
        return {
            "file": self.file_path,
            "score": self.score,
            "priority": self.priority,
            "roles": [r.value for r in self.roles],
            "reason": self.reason,
            "confidence": self.confidence,
        }


class PriorityEngine:
    """Assigns priority buckets to files based on role scores."""
    
    @staticmethod
    def assign_priority(score: float) -> str:
        """
        Assign priority bucket based on score.
        
        Args:
            score: Relevance score (0.0 - 1.0)
            
        Returns:
            "primary", "supporting", "context", or "hidden"
        """
        for priority_name, (min_val, max_val) in PRIORITY_THRESHOLDS.items():
            if min_val <= score <= max_val:
                return priority_name
        return "hidden"
    
    @staticmethod
    def create_role_view(
        role: Role,
        file_scores: Dict[str, Dict[Role, float]],
        dependencies: Dict[str, Set[str]] = None,
    ) -> Dict[str, List[PrioritizedFile]]:
        """
        Create a prioritized view for a specific role.
        
        Args:
            role: The role to create view for
            file_scores: {"file_path": {"role1": score1, ...}}
            dependencies: {"file_path": {"dependency1", "dependency2", ...}}
            
        Returns:
            {
                "primary": [PrioritizedFile, ...],
                "supporting": [PrioritizedFile, ...],
                "context": [PrioritizedFile, ...],
            }
        """
        view = {
            "primary": [],
            "supporting": [],
            "context": [],
        }
        
        # Track which files we've already added
        added = set()
        
        # 1. Score files directly for this role
        scored_files = []
        for file_path, scores in file_scores.items():
            score = scores.get(role, 0.0)
            if score > 0.0:
                priority = PriorityEngine.assign_priority(score)
                roles = {r for r, s in scores.items() if s > 0.3}
                scored_files.append((file_path, score, priority, roles))
        
        # 2. Add direct matches
        scored_files.sort(key=lambda x: x[1], reverse=True)
        for file_path, score, priority, roles in scored_files:
            # Skip hidden priority files (they're not shown by default)
            if priority == 'hidden':
                continue
            
            pf = PrioritizedFile(
                file_path=file_path,
                score=score,
                priority=priority,
                roles=roles,
            )
            view[priority].append(pf)
            added.add(file_path)
        
        # 3. Add cross-role context (dependencies)
        if dependencies:
            context_additions = PriorityEngine._get_cross_role_context(
                role, file_scores, dependencies, added
            )
            for pf in context_additions:
                # Skip hidden priority files
                if pf.priority == 'hidden':
                    continue
                view[pf.priority].append(pf)
                added.add(pf.file_path)
        
        # Limit results to avoid overwhelming UI
        for priority in view:
            view[priority] = view[priority][:MAX_FILES_PER_PRIORITY]
        
        return view
    
    @staticmethod
    def _get_cross_role_context(
        role: Role,
        file_scores: Dict[str, Dict[Role, float]],
        dependencies: Dict[str, Set[str]],
        already_added: Set[str],
    ) -> List[PrioritizedFile]:
        """
        Get files that should be shown as context due to dependencies.
        
        Example: If Backend role is selected, show Frontend files that call
        backend APIs as CONTEXT.
        """
        context_files = []
        
        # Cross-role patterns
        cross_role_context = {
            Role.FRONTEND: {
                Role.BACKEND: ("API calls", 0.35),  # Show backend APIs frontend uses
                Role.SECURITY: ("Auth concerns", 0.2),
            },
            Role.BACKEND: {
                Role.FRONTEND: ("UI usage", 0.25),  # Show frontend components
                Role.DATA: ("Data access", 0.5),
                Role.SECURITY: ("Auth/validation", 0.4),
            },
            Role.DEVOPS: {
                Role.BACKEND: ("Service deployment", 0.3),
                Role.FRONTEND: ("CDN/static hosting", 0.2),
            },
            Role.DATA: {
                Role.BACKEND: ("Data consumers", 0.4),
                Role.AI_ML: ("Pipeline input", 0.5),
            },
        }
        
        if role not in cross_role_context:
            return context_files
        
        target_roles = cross_role_context[role]
        
        for file_path, scores in file_scores.items():
            if file_path in already_added:
                continue
            
            for target_role, (reason, boost_score) in target_roles.items():
                if scores.get(target_role, 0.0) > 0.5:
                    priority = PriorityEngine.assign_priority(boost_score)
                    if priority in ["supporting", "context"]:
                        roles = {r for r, s in scores.items() if s > 0.3}
                        context_files.append(PrioritizedFile(
                            file_path=file_path,
                            score=boost_score,
                            priority=priority,
                            roles=roles,
                            reason=reason,
                        ))
        
        # Sort by score and deduplicate
        context_files.sort(key=lambda x: x.score, reverse=True)
        seen = {pf.file_path for pf in context_files}
        return [pf for pf in context_files if pf.file_path not in already_added]
    
    @staticmethod
    def merge_multi_role_view(
        roles: List[Role],
        file_scores: Dict[str, Dict[Role, float]],
        dependencies: Dict[str, Set[str]] = None,
        merge_strategy: str = "max",  # "max", "average", "weighted"
    ) -> Dict[str, List[PrioritizedFile]]:
        """
        Merge views from multiple roles.
        
        Args:
            roles: List of roles to merge
            file_scores: File scores by role
            dependencies: Dependency graph
            merge_strategy: How to combine scores
            
        Returns:
            Merged prioritized view
        """
        # Compute merged scores
        merged_scores = {}
        
        for file_path, scores in file_scores.items():
            role_scores = [scores.get(r, 0.0) for r in roles]
            
            if merge_strategy == "max":
                merged_score = max(role_scores)
            elif merge_strategy == "average":
                merged_score = sum(role_scores) / len(roles)
            elif merge_strategy == "weighted":
                # Weight by number of relevant roles
                merged_score = sum(role_scores) / len(roles) * (len([s for s in role_scores if s > 0]) / len(roles))
            else:
                merged_score = max(role_scores)
            
            merged_scores[file_path] = merged_score
        
        # Create view using merged scores
        view = {
            "primary": [],
            "supporting": [],
            "context": [],
        }
        
        scored_files = [
            (file_path, score)
            for file_path, score in merged_scores.items()
        ]
        scored_files.sort(key=lambda x: x[1], reverse=True)
        
        for file_path, score in scored_files:
            priority = PriorityEngine.assign_priority(score)
            # Skip hidden priority files
            if priority == 'hidden':
                continue
            
            relevant_roles = {r for r in roles if file_scores[file_path].get(r, 0.0) > 0.3}
            
            pf = PrioritizedFile(
                file_path=file_path,
                score=score,
                priority=priority,
                roles=relevant_roles,
            )
            view[priority].append(pf)
        
        # Limit results
        for priority in view:
            view[priority] = view[priority][:MAX_FILES_PER_PRIORITY]
        
        return view


def assign_priority(score: float) -> str:
    """
    Convenience function: assign priority based on score.
    
    Example:
        priority = assign_priority(0.85)  # "primary"
    """
    return PriorityEngine.assign_priority(score)

File: src/priorityEngine.ts
/**
 * Priority assignment and bucketing engine.
 *
 * Converts role scores into priority buckets:
 * - PRIMARY (0.7 – 1.0): Must understand
 * - SUPPORTING (0.4 – 0.7): Related logic
 * - CONTEXT (0.1 – 0.4): Additional context
 * - HIDDEN (0.0 – 0.1): Not shown by default
 */

import { Role, PRIORITY_THRESHOLDS, MAX_FILES_PER_PRIORITY } from "./config";
import { PrioritizedFile, RoleView } from "./types";

export class PrioritizedFileImpl implements PrioritizedFile {
  filePath: string;
  score: number;
  priority: "primary" | "supporting" | "context" | "hidden";
  roles: Set<Role>;
  reason?: string;
  confidence?: number;
  explanation?: string;
  hasDependencies?: boolean;
  hasDependents?: boolean;

  constructor(
    filePath: string,
    score: number,
    priority: "primary" | "supporting" | "context" | "hidden",
    roles: Set<Role>
  ) {
    this.filePath = filePath;
    this.score = score;
    this.priority = priority;
    this.roles = roles;
  }

  // Alias for filePath for compatibility
  get path(): string {
    return this.filePath;
  }

  toDict(): Record<string, any> {
    return {
      file: this.filePath,
      path: this.filePath,
      score: this.score,
      priority: this.priority,
      roles: Array.from(this.roles).map((r) => r.toString()),
      reason: this.reason || "",
      confidence: this.confidence || 1.0,
      explanation: this.explanation || "",
      hasDependencies: this.hasDependencies || false,
      hasDependents: this.hasDependents || false,
    };
  }
}

export class PriorityEngine {
  /**
   * Assign priority bucket based on score.
   */
  static assignPriority(
    score: number
  ): "primary" | "supporting" | "context" | "hidden" {
    for (const [priority, [minVal, maxVal]] of Object.entries(PRIORITY_THRESHOLDS)) {
      if (score >= minVal && score <= maxVal) {
        return priority as "primary" | "supporting" | "context" | "hidden";
      }
    }
    return "hidden";
  }

  /**
   * Create prioritized view for a specific role.
   */
  static createRoleView(
    role: Role,
    fileScores: Record<string, Record<Role, number>>,
    dependencies?: Record<string, Set<string>>
  ): RoleView {
    const view: RoleView = {
      primary: [],
      supporting: [],
      context: [],
    };

    // Score files for this role
    const scoredFiles: Array<[string, number, Set<Role>]> = [];

    for (const [filePath, scores] of Object.entries(fileScores)) {
      const roleScore = scores[role] || 0.0;

      if (roleScore > 0.0) {
        // Get all roles this file is relevant to (score > 0.3)
        const relevantRoles = new Set<Role>();
        for (const [r, score] of Object.entries(scores)) {
          if (score > 0.3) {
            relevantRoles.add(r as Role);
          }
        }

        scoredFiles.push([filePath, roleScore, relevantRoles]);
      }
    }

    // Sort by score descending
    scoredFiles.sort(([, a], [, b]) => b - a);

    // Assign to priority buckets
    for (const [filePath, score, roles] of scoredFiles) {
      const priority = PriorityEngine.assignPriority(score);

      if (priority !== "hidden") {
        const pf = new PrioritizedFileImpl(filePath, score, priority, roles);

        if (dependencies) {
          pf.hasDependencies = Boolean(dependencies[filePath]?.size);
        }

        view[priority].push(pf);
      }
    }

    // Limit results
    for (const priority of ["primary", "supporting", "context"] as const) {
      view[priority] = view[priority].slice(0, MAX_FILES_PER_PRIORITY);
    }

    return view;
  }

  /**
   * Merge views from multiple roles.
   */
  static mergeMultiRoleView(
    roles: Role[],
    fileScores: Record<string, Record<Role, number>>,
    dependencies?: Record<string, Set<string>>,
    mergeStrategy: "max" | "average" | "weighted" = "max"
  ): RoleView {
    const view: RoleView = {
      primary: [],
      supporting: [],
      context: [],
    };

    // Compute merged scores
    const mergedScores: Record<string, number> = {};

    for (const [filePath, scores] of Object.entries(fileScores)) {
      const roleScores = roles.map((r) => scores[r] || 0.0);

      let mergedScore: number;

      if (mergeStrategy === "max") {
        mergedScore = Math.max(...roleScores);
      } else if (mergeStrategy === "average") {
        mergedScore = roleScores.reduce((a, b) => a + b, 0) / roles.length;
      } else {
        // weighted
        const relevantCount = roleScores.filter((s) => s > 0).length;
        mergedScore =
          (roleScores.reduce((a, b) => a + b, 0) / roles.length) *
          (relevantCount / roles.length);
      }

      mergedScores[filePath] = mergedScore;
    }

    // Create view using merged scores
    const scoredFiles: Array<[string, number]> = [];

    for (const [filePath, score] of Object.entries(mergedScores)) {
      if (score > 0.0) {
        scoredFiles.push([filePath, score]);
      }
    }

    scoredFiles.sort(([, a], [, b]) => b - a);

    // Assign to priority buckets
    for (const [filePath, score] of scoredFiles) {
      const priority = PriorityEngine.assignPriority(score);

      if (priority !== "hidden") {
        // Get roles this file is relevant to
        const relevantRoles = new Set<Role>();
        for (const r of roles) {
          if (fileScores[filePath][r] > 0.3) {
            relevantRoles.add(r);
          }
        }

        const pf = new PrioritizedFileImpl(filePath, score, priority, relevantRoles);

        if (dependencies) {
          pf.hasDependencies = Boolean(dependencies[filePath]?.size);
        }

        view[priority].push(pf);
      }
    }

    // Limit results
    for (const priority of ["primary", "supporting", "context"] as const) {
      view[priority] = view[priority].slice(0, MAX_FILES_PER_PRIORITY);
    }

    return view;
  }
}
